File: whiteboard/app/src/main/java/com/ditto/whiteboard/data/DebugTransportSettings.kt
package com.ditto.whiteboard.data

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit

/**
 * Debug-only transport overrides applied at Ditto construction. Toggling any of these requires an
 * app restart to take effect; the Troubleshooting screen surfaces them for protocol testing.
 */
class DebugTransportSettings(context: Context) {
  private val prefs: SharedPreferences =
    context.applicationContext.getSharedPreferences("whiteboard_debug_transports", Context.MODE_PRIVATE)

  init {
    // Before v2, the default disabled ordinary IP-multicast *discovery* even while LAN was
    // enabled. Correct that legacy default once; later user changes remain authoritative.
    if (!prefs.getBoolean(KEY_LAN_DISCOVERY_DEFAULT_V2, false)) {
      prefs.edit {
        putBoolean(KEY_MULTICAST, true)
        putBoolean(KEY_LAN_DISCOVERY_DEFAULT_V2, true)
      }
    }
  }

  var bluetoothLeEnabled: Boolean
    get() = prefs.getBoolean(KEY_BLE, true)
    set(value) {
      prefs.edit { putBoolean(KEY_BLE, value) }
    }

  var wifiAwareEnabled: Boolean
    get() = prefs.getBoolean(KEY_WIFI_AWARE, true)
    set(value) {
      prefs.edit { putBoolean(KEY_WIFI_AWARE, value) }
    }

  var lanEnabled: Boolean
    get() = prefs.getBoolean(KEY_LAN, true)
    set(value) {
      prefs.edit { putBoolean(KEY_LAN, value) }
    }

  var mdnsEnabled: Boolean
    get() = prefs.getBoolean(KEY_MDNS, true)
    set(value) {
      prefs.edit { putBoolean(KEY_MDNS, value) }
    }

  var multicastEnabled: Boolean
    get() = prefs.getBoolean(KEY_MULTICAST, true)
    set(value) {
      prefs.edit { putBoolean(KEY_MULTICAST, value) }
    }

  companion object {
    private const val KEY_BLE = "bluetooth_le_enabled"
    private const val KEY_WIFI_AWARE = "wifi_aware_enabled"
    private const val KEY_LAN = "lan_enabled"
    private const val KEY_MDNS = "mdns_enabled"
    private const val KEY_MULTICAST = "multicast_enabled"
    private const val KEY_LAN_DISCOVERY_DEFAULT_V2 = "lan_discovery_default_v2"
  }
}

File: whiteboard-apple/Whiteboard/App/DebugTransportSettings.swift
import Foundation

/// Debug-only transport overrides applied at Ditto construction. Toggling any of these requires an
/// app restart to take effect; the Troubleshooting screen surfaces them for protocol testing.
struct DebugTransportSettings: Sendable {
  var bluetoothLEEnabled: Bool
  var lanEnabled: Bool
  var mdnsEnabled: Bool
  var multicastEnabled: Bool
  var awdlEnabled: Bool

  static let `default` = DebugTransportSettings(
    bluetoothLEEnabled: true,
    lanEnabled: true,
    mdnsEnabled: true,
    multicastEnabled: true,
    awdlEnabled: true
  )

  private enum Key {
    static let ble = "debugTransports.bluetoothLE"
    static let lan = "debugTransports.lan"
    static let mdns = "debugTransports.mdns"
    static let multicast = "debugTransports.multicast"
    static let lanDiscoveryDefaultV2 = "debugTransports.lanDiscoveryDefaultV2"
    static let awdl = "debugTransports.awdl"
  }

  static func load(from defaults: UserDefaults = .standard) -> DebugTransportSettings {
    // Before v2, the default disabled ordinary IP-multicast *discovery* even while LAN was
    // enabled. Correct that legacy default once; users can still turn the debug setting off
    // afterward and their explicit choice is retained.
    if !defaults.bool(forKey: Key.lanDiscoveryDefaultV2) {
      defaults.set(true, forKey: Key.multicast)
      defaults.set(true, forKey: Key.lanDiscoveryDefaultV2)
    }
    func read(_ key: String, fallback: Bool) -> Bool {
      defaults.object(forKey: key) as? Bool ?? fallback
    }
    return DebugTransportSettings(
      bluetoothLEEnabled: read(Key.ble, fallback: true),
      lanEnabled: read(Key.lan, fallback: true),
      mdnsEnabled: read(Key.mdns, fallback: true),
      multicastEnabled: read(Key.multicast, fallback: true),
      awdlEnabled: read(Key.awdl, fallback: true)
    )
  }

  func save(to defaults: UserDefaults = .standard) {
    defaults.set(bluetoothLEEnabled, forKey: Key.ble)
    defaults.set(lanEnabled, forKey: Key.lan)
    defaults.set(mdnsEnabled, forKey: Key.mdns)
    defaults.set(multicastEnabled, forKey: Key.multicast)
    defaults.set(awdlEnabled, forKey: Key.awdl)
  }
}
